Activity Log: break timeline before the Rewind confirmation dialog

eliorivero opened the pull request:
Fixes #18745

#### Before
<img width="403" alt="captura de pantalla 2017-10-26 a la s 22 34 19" src="https://user-images.githubusercontent.com/1041600/32084279-807f0b6c-ba9e-11e7-909b-5c6799b7e5b5.png">

#### After
<img width="397" alt="captura de pantalla 2017-10-26 a la s 22 33 56" src="https://user-images.githubusercontent.com/1041600/32084186-01201a82-ba9e-11e7-8f25-da247676e24e.png">

#### Before
<img width="730" alt="captura de pantalla 2017-10-27 a la s 18 53 58" src="https://user-images.githubusercontent.com/1041600/32126991-311087ca-bb4a-11e7-866b-19f15d228aee.png">

#### After
<img width="629" alt="captura de pantalla 2017-10-27 a la s 19 02 46" src="https://user-images.githubusercontent.com/1041600/32126994-344600f0-bb4a-11e7-9cac-4322eac8cbd9.png">

dmsnell commented on client/my-sites/stats/activity-log-day/index.jsx:
```diff
@@ -177,6 +177,16 @@ class ActivityLogDay extends Component {
+	 * @returns {boolean} True if this is the React element for the Rewind dialog.
+	 */
+	static isRewindDialog( item ) {
+		return 'activity-rewind-dialog' === get( item, 'key', '' );
```
note that the `get()` here is unnecessary if `item` is defined, which it looks like it will be, because if `item.key` doesn't exist or is _anything_ other than the string we're comparing it to, then the equality will fail…

also, is there a reason we made this `static` here? typically we would put this outside of the React component and into the module scope, which should end up leading to a smaller/faster build, as every method inside of the component is generated for every instance of the component while module-level functions are just created once.

performance shouldn't matter in most cases, but on the other hand it's so trivial to move them out and so awkward to see a `static` definition of a function in JavaScript/React…

```js
const isRewindDialog = ( { key } ) => key === 'activity-rewind-dialog'

class ActivityLogDay extends Component {
	…
	log => isRewindDialog( log )…
	…
}
```

dmsnell commented on client/my-sites/stats/activity-log-day/index.jsx:
```diff
@@ -224,18 +234,36 @@ class ActivityLogDay extends Component {
+										siteId={ siteId }
+									/>
+								)
+						) }
```
This refactoring is a bit confusing to me: do you think you could try and explain in writing the objective of the code?

It _appears_ as though the goal is to find out which log items occur immediately _before_ a rewind dialog.

The code itself stands out to me for a few reasons:
  - it's unexpected to see `map( flatMap() )` because I think that our goal is do end up with what `flatMap()` is producing, so I'm not sure why `map()` is there (not to say it can't belong there)
  - the use of `collection[ index + 1 ]` is a warning sign because we know that this will end up returning a non-existent element at the end of every spin through the loop - we're intentionally encoding array-out-of-bounds errors
  - the nested transforms are just a bit hard for me to follow

What I think I see happening here is a second loop after an initial loop which is trying to find those items which have a rewind dialog open, predicting from earlier iteration and separating them out. I wonder if we couldn't extract this processing to the data structure before we get into the JSX and end up with one iteration or at least processing of the data which lets us go once through it in React…

dmsnell replied:
I'd like to raise the challenge again to think about the data, data structures, and transform operations we are doing _to_ this data so that we can hone in on it in such a way that the logic remains clear and separated from the UI and so that the UI just kind of "falls into place" naturally since our data is prepared for the visual goals.

here's just one thing I'm brainstorming, and I think the assumption it's based on holds - we'll only have one open restore dialog.

> we want to render each item in a list, but if an item has been selected for a restore, then we should display a restore dialog before that item, and the timeline above that restore dialog ought to break

from this specification I suspect we could accomplish this by splitting our array and treating the two halves differently. we could actualize the same specs with a reducer function as well, carrying forward state through its accumulator (to track things like _the last item_ and _the running list of items_) but I think in this case we can get by with a simpler setup which requires no state tracking and no index tracking…

**we have a list of items and up to one of them may have an open restore dialog**

this looks like `_______-____`

we can split this into a _prefix_ list, the possibly open item, and the _postfix_ list. actually, since all we need to do is operate at the first boundary, we can see it as a first half and possibly-empty second half of a list

now we see this with our added `+` dialog - `_____+------` because everything _after_ the rewind dialog is normal

```js
import { findIndex, initial, last } from 'lodash'

const splitByRewind = ( logs, activityId ) => {
	const rewindIndex = findIndex( logs, log => log.activityId === activityId );

	// no dialog is open, just return the front list
	if ( -1 === rewindIndex ) {
		return [ logs, [] ];
	}

	return [
		logs.slice( 0, rewindIndex ), // before the rewind
		logs.slice( rewindIndex ), // at and after rewind
	];
}
```

now down somewhere in our render function we can add in the visual bits…

```js
const [ beforeRewind, afterRewind ] = splitByRewind( rewriteStream( logs ), requestedRestoreActivityId )

…

const LogItem = ( { log, extraClasses } ) => (
	<ActivityLogItem
		className={ extraClasses }
		applySiteOffset={ applySiteOffset }
		disableRestore={ disableRestore }
		hideRestore={ hideRestore }
		key={ log.activityId }
		log={ log }
		requestRestore={ requestRestore }
		siteId={ siteId }
	/>
);

{ /* insert the rewind dialog before an open rewind event and stylize the preceding item */ }
{ initial( beforeRewind ).map( log => <LogItem { ...{ log } } /> }
{ last( beforeRewind ) && <LogItem { ...{ log: last( beforeRewind ), extraClasses: 'is-before-rewind' } } /> }
{ afterRewind.length && rewindConfirmDialog }
{ afterRewind.map( log => <LogItem { ...{ log } } /> }
```

so in some senses we are losing some context by making a local custom `<LogItem />` but what we gain is a more straight-forward declaration of how the UI will look without mixing in the logic of separating and finding the predecessor to the rewind.

could we have done this differently? maybe it would have made it easier still in the render to have returned all of these three distinct places (normal items, the immediate predecessor to the rewind, the items after the rewind dialog) as arrays which would let us take advantage of the safe-defaulting empty state of lists…

```js
// returns [ before, rightBefore, after ]
const splitByRewind = ( logs, activityId ) => {
	const rewindIndex = findIndex( logs, log => log.activityId === activityId );

	// no dialog is open
	if ( -1 === rewindIndex ) {
		return [ logs, [], [], ];
	}

	// dialog is open at first item
	if ( 0 === rewindIndex ) {
		return [ [], [], logs ];
	}

	// first item is immediately before
	if ( 1 === rewindIndex ) {
		return [ [], [ logs[ 0 ] ], logs.slice( 1 ) ];
	}

	// everything else is standard
	return [
		logs.slice( 0, rewindIndex - 1 ),
		[ logs[ rewindIndex - 1 ] ],
		logs.slice( rewindIndex ),
	];
}

…

render() {
	…
	const [ newer, ontop, older ] = splitByRewind( rewriteStream( logs ), requestedRestoreActivityId );
	…
		{ newer.map( log => <LogItem { ...{ log } } /> ) }
		{ ontop.map( log => <LogItem { ...{ log, extraClasses: 'is-before-rewind' } } /> ) }
		{ older && rewindConfirmDialog }
		{ older.map( log => <LogItem { ...{ log } } /> ) }
	…
}
```

hope this gives you some ideas to work with! please feel free to poke me about it

eliorivero commented:
The goal of all this is to add a class to the element found right before the Rewind dialog. So far the `flatMap` takes the function that creates the tuple with an event and possibly the Rewind dialog. If it's not there, it will add false to the tuple. All those false elements are added to the list. This is a side effect of this new code, that it won't add those tuples with false, with inherently less items and more performant. However, the real motivation behind this code, is that in the previous list, we couldn't know the element previous to the Rewind dialog since the tuple is created with the current event. So I needed to find out the previous event, and I processed through `map` again to find out if the next item to insert is the Rewind dialog.

eliorivero commented:
Thanks for the review Dennis, I've moved the static method outside for better performance, thanks for pointing this out, and also introduced a check to avoid checking the array item beyond its limits.

I also didn't like when I saw `flatMap` was used to iterate over the elements since it's adding all these tuples with `false` in it and leaving the JSX parser to exclude them when it's rendered. I've removed it now and instead use `splice` to insert the element only when it's supposed to be inserted, instead of checking every time like the approach `flatMap` was doing.

For now, I won't go into refactoring this any further since I have some other tasks that need attention. I understand that you've the best intentions suggesting these refactors, and also you're usually (always?) right. While I'd love to introduce them, as much as I enjoy functional programming too, we have some deadlines to meet and right now we shouldn't be derailed from meeting them. If you have the time, feel free to introduce them in separate PRs, will be happy to review and test. 
On the other hand, comments like moving the static method outside for better performance, or avoiding checking the array beyond limits (which resulted in removing the `get` function) or complaining about running `map` on the result of `flatMap` are awesome, and have a good balance of making things right and moving forward to meet our deadlines.

MichaelArestad commented:
I tried testing this without success. Possibly because there are no more three dot menus on my test site to restore to specific events. I can restore to the day, though. ¯\_(ツ)_/¯

eliorivero commented:
Try doing some actions on your site, install a plugin, wait for 3 min, activate it, wait for 3, deactivate and delete it, wait 3 minutes again (the 3 minutes is purely based in my xp). You'll then get some more events and ellipsis menus in the day.
